front/src/modules/timesStops/helpers/computeMargins.tsx

```diff
@@ -0,0 +1,88 @@
+import type { SimulationResponseSuccess } from 'applications/operationalStudies/types';
+import type { TrainScheduleResult } from 'common/api/generatedEditoastApi';
+import type { OperationalPointWithTimeAndSpeed } from 'modules/trainschedule/components/DriverTrainScheduleV2/types';
+import { interpolateValue } from 'modules/trainschedule/components/DriverTrainScheduleV2/utils';
+import { mToMm, msToS } from 'utils/physics';
+
+import { formatDigitsAndUnit } from './utils';
+
+const emptyMargins = {
+  theoreticalMargin: '',
+  theoreticalMarginSeconds: '',
+  calculatedMarginSeconds: '',
+  diffMarginsSeconds: '',
+};
+
+function getTheoreticalMargin(selectedTrainSchedule: TrainScheduleResult, pathStepId: string) {
+  const pathStepIndex = selectedTrainSchedule.path.findIndex(
+    (pathStep) => pathStep.id === pathStepId
+  );
+  let theoreticalMarginIndex;
+  if (pathStepIndex === 0) {
+    theoreticalMarginIndex = 0;
+  } else {
+    const theoriticalMarginBoundaryIndex = selectedTrainSchedule.margins?.boundaries?.findIndex(
+      (id) => id === pathStepId
+    );
+    if (
+      theoriticalMarginBoundaryIndex === undefined ||
+      theoriticalMarginBoundaryIndex < 0 ||
+      theoriticalMarginBoundaryIndex > selectedTrainSchedule.margins!.values.length - 2
+    ) {
+      return undefined;
+    }
+    theoreticalMarginIndex = theoriticalMarginBoundaryIndex + 1;
+  }
+  return selectedTrainSchedule.margins!.values[theoreticalMarginIndex];
+}
+
+function computeDuration(
+  trainSimulation: SimulationResponseSuccess,
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  key: 'base' | 'provisional' | 'final_output'
+) {
+  const opPointTime = interpolateValue(trainSimulation[key], mToMm(opPoint.position), 'times');
+  const nextOpPointTime = interpolateValue(
+    trainSimulation[key],
+    mToMm(nextOpPoint.position),
+    'times'
+  );
+  return msToS(nextOpPointTime - opPointTime);
+}
+
+function computeMargins(
+  trainSimulation: SimulationResponseSuccess,
+  opPoint: OperationalPointWithTimeAndSpeed,
+  nextOpPoint: OperationalPointWithTimeAndSpeed,
+  selectedTrainSchedule: TrainScheduleResult,
+  pathStepId: string
+) {
+  const theoreticalMargin = getTheoreticalMargin(selectedTrainSchedule, pathStepId);
+
+  // durations to go from current scheduled point to next scheduled point.
+  // base = no margin
+  // provisional = margins
+  // final = margin + requested arrival times
+  const baseDuration = computeDuration(trainSimulation, opPoint, nextOpPoint, 'base');
```

Comment:
And what if you change the signature of the computeDuration function like this (use ReportTrainV2 directly) ?
```
function computeDuration(
  trainSimulation: ReportTrainV2,
  opPoint: OperationalPointWithTimeAndSpeed,
  nextOpPoint: OperationalPointWithTimeAndSpeed
) {}
```
and then:
```suggestion
  const baseDuration = computeDuration(trainSimulation.base, opPoint, nextOpPoint);
```

Reply:
good stuff, done